staging/src/k8s.io/apimachinery/pkg/util/waitgroup/optionalwaitgroup.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package waitgroup
+
+import (
+	"sync"
+)
+
+// OptionalWaitGroup is a WaitGroup or nothing.  Use it to track when
+// goroutines start or stop or begin or end waiting if the pointer is
+// not nil.  This abstraction is intended to make it easier to write a
+// package that can be unit tested by a function that advances a fake
+// clock once all the available work is done.  Real clients set the
+// pointer to nil because they do not need this tracking.
+type OptionalWaitGroup struct {
+	WG *sync.WaitGroup
+}
+
+// NoWaitGroup constructs an OptionalWaitGroup with a nil pointer
+func NoWaitGroup() OptionalWaitGroup {
+	return OptionalWaitGroup{WG: nil}
+}
+
+// WrapWaitGroupPointer is the abstraction function
+func WrapWaitGroupPointer(wg *sync.WaitGroup) OptionalWaitGroup {
+	return OptionalWaitGroup{WG: wg}
+}
+
+// Increment adds one to the count of active goroutines if and only if the pointer is not nil
+func (owg OptionalWaitGroup) Increment() {
+	if owg.WG != nil {
+		owg.WG.Add(1)
+	}
+}
+
+// Decrement subtracts one from the count of active goroutines if and only if the pointer is not nil
+func (owg OptionalWaitGroup) Decrement() {
+	if owg.WG != nil {
+		owg.WG.Add(-1)
```

Comment:
This is not a normal thing to do to a waitgroup, generally we call Done(). Are you sure a wait group is the right abstraction? It is unsafe to call Add() after someone has called Wait().

Reply:
`Done()` is just short-hand for `Add(-1)`, right?

Re your remark "It is unsafe to call Add() after someone has called Wait()": are you referring to the remark in https://golang.org/pkg/sync/#WaitGroup.Add that "If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned"?  The latter is a more precise statement, and the usage meets that restriction.

Yes, I think WaitGroup is the right abstraction.  It's job is to keep track of how many goroutines in some category are active.  Probably what is missing is enough commentary to document what is really going on.  And there is the issue of where to write the documentation.  This whole file is just a utility for code that (a) optionally uses a WaitGroup to track _all_ the goroutines relevant to that code and (b) is tested with a FakeEventClock.  Regular clients do not need the tracking and so give this code `nil` rather than a non-nil pointer to a WaitGroup.  The unit test code, OTOH, constructs a FakeEventClock and gives it to the code under test; both that code and the FakeEventClock get the same pointer to an actual WaitGroup.  The code under test updates the WaitGroup iff the pointer is not nil.  The FakeEventClock is the only thing that calls Wait on that WaitGroup, and Wait is used only in FakeEventClock::Run, and the test code does not have concurrent calls to Run, and the test code does not update the WaitGroup concurrently with the Wait.

Reply:
On second thought, maybe we _should_ introduce an abstraction here.  The point would be that the abstraction _only_ gives access to Add/Done, because that is all that the code under test needs.  The abstraction might look something like the following.

```
// GoRoutineCounter keeps track of the number of active goroutines
// working on/for something.  This is a utility that makes such code more
// testable.  The code uses this utility to report the number of active
// goroutines to the test code, so that the test code can advance a fake
// clock when and only when the code being tested has finished all
// the work that is ready to do at a given time.
type GoRoutineCounter interface {
    // Add adds the given delta to the count of active goroutines.
    // Call Add(1) before forking a goroutine, Add(-1) when that goroutine finishes.
    // Call Add(-1) just before waiting on something from another goroutine (e.g.,
    // just before a `select`), and Add(1) just after that something arrives.
    Add(delta int)
}
```

Reply:
Thinking more about how to use the type system to illuminate/enforce the intended usage, I think we could take the following approach.  Change `NewFakeEventClock` so that instead of _taking_ a GoRoutineCounter as a paramter, that function _returns_ a freshly-constructed GoRoutineCounter based on a private implementation type.  That way the clock code can call Wait on the underlying WaitGroup but other code can not (assuming we move the clock code to its own package).

Reply:
I've removed `optionalwaitgroup.go` and added the suggested GoRoutineCounter interface into `fairqueuing/interfaces.go` wiring it through.

Reply:
Does it make sense for NewFakeEventClock to return more than the FakeEventClock?  I will update my implementation is so (I have not made that change yet).

Reply:
Updated the code to include the GoRoutineCounter and NewFakeEventClock changes, resolving the thread.